MRG: Resampling without losing events

wmvanvliet opened the pull request:
Resampling the stim channel is a tricky business. If you are not
careful, you might lose events.

Consider the following stim data, all zeros with two event onsets:

      y = [0, 0, 1, 0, 0, 0, 0, 1, 0, 0]

If we naively downsample by a factor of 2, this becomes:

     y = [0, 1, 0, 0, 0]

whoops, event lost. In the current implementation, no warning is given whatsoever.

This commit first extracts the onsets on which the stim signal changes:

     onsets = [0, 2, 3, 6, 7]

resample those:

    onsets = [0, 1, 1, 3, 3]

nudge overlapping onsets (by default by maximally 1 sample, but this can be changed):

    onsets = [0, 1, 2, 3, 4]

and reconstruct the resampled stim data:

    y = [0, 1, 0, 1, 0]

no events lost!

If events get delayed too much or fall out of bounds, it fails loud.

mshamalainen commented:
Looking at the mne_browse_raw C code may help in solving this problem. -  Matti

> On Jul 14, 2015, at 9:20 PM, Marijn van Vliet <notifications@github.com> wrote:
> 
> Resampling the stim channel is a tricky business. If you are not
> careful, you might lose events.
> 
> Consider the following stim data, all zeros with two event onsets:
> 
>   y = [0, 0, 1, 0, 0, 0, 0, 1, 0, 0]
> If we naively downsample by a factor of 2, this becomes:
> 
>  y = [0, 1, 0, 0, 0]
> whoops, event lost.
> 
> This commit first extracts the onsets on which the stim signal changes:
> 
>  onsets = [0, 2, 3, 6, 7]
> resample those:
> 
> onsets = [0, 1, 1, 3, 3]
> nudge overlapping onsets (by default by maximally 1 sample, but this can be changed):
> 
> onsets = [0, 1, 2, 3, 4]
> and reconstruct the resampled stim data:
> 
> y = [0, 1, 0, 1, 0]
> no events lost!
> 
> If events get delayed too much or fall out of bounds, it fails hard.
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/mne-tools/mne-python/pull/2298 <https://github.com/mne-tools/mne-python/pull/2298>
> Commit Summary
> 
> Resampling without losing events
> File Changes
> 
> M mne/filter.py <https://github.com/mne-tools/mne-python/pull/2298/files#diff-0> (65)
> M mne/io/base.py <https://github.com/mne-tools/mne-python/pull/2298/files#diff-1> (36)
> M mne/tests/test_filter.py <https://github.com/mne-tools/mne-python/pull/2298/files#diff-2> (52)
> Patch Links:
> 
> https://github.com/mne-tools/mne-python/pull/2298.patch <https://github.com/mne-tools/mne-python/pull/2298.patch>
> https://github.com/mne-tools/mne-python/pull/2298.diff <https://github.com/mne-tools/mne-python/pull/2298.diff>
> —
> Reply to this email directly or view it on GitHub <https://github.com/mne-tools/mne-python/pull/2298>.
> 



The information in this e-mail is intended only for the person to whom it is
addressed. If you believe this e-mail was sent to you in error and the e-mail
contains patient information, please contact the Partners Compliance HelpLine at
http://www.partners.org/complianceline . If the e-mail was sent to you in error
but does not contain patient information, please contact the sender and properly
dispose of the e-mail.

wmvanvliet commented:
This PR solves the problem completely. Now researchers can finally sleep soundly at night knowing no events are ever dropped again if it can be helped at all.

dgwakeman commented:
Well, it may also introduce a problem, I can think of at least one case where could be problematic:
On many Neuromag systems the trigger is not always correct on the first sample of a new trigger. This is why we have the shortest_event field on find_events (which defaults to 2). #1102
I know that @mshamalainen included some neat tricks in the C-code, which we may want to look at for taking care of these edge cases where this code might exacerbate the issue.

agramfort commented:
@wmvanvliet I've given you access to the C code so you can have a look
at the hacks of matti :)

wmvanvliet commented:
thanks @mshamalainen and @agramfort, I will take a look at the C code

wmvanvliet commented:
@dgwakeman so in (very) rare cases this Neuromag bug can also cause event codes to *change* due to naive resampling. For example:

    y = [0, 0, 0, 1, 2, 2, 0, 0, 0, 0, 0, 0]

downsample with a factor of 3:

    y = [0, 1, 0, 0]

The implementation in this PR would regard the `1, 2, 2` bit as two events (1 and 2) and make sure to retain them both:

    y = [0, 1, 2, 0]

Arguably better... although this would be a very rare situation.
Users of Neuromag should be very careful in any case.

Alternatively, how about a `resample_events` function? The user could first find events on the non-resampled data, using whatever hacks and tricks necessary. Then resample the data and event matrices separately. That would make the code a little bit easier as well.

dgwakeman commented:
Is this the only tricky thing you found in the C-code. I don't have access myself: plus I'm a bit dense, so I don't know if I would be able to read it that well. ;)

I didn't have any strong design ideas (IMO: one sample triggers are the real problem). I just didn't want to see the code around this area getting too complex and developing strange interactions with the defaults (e.g. your code is more likely to output single sample triggers; which would be ignored by the defaults in find_events), but I'm not blocking. I just wanted to voice some concerns. A resample events option might be good to have. The usage pattern could look like:

events = raw.find_events;
raw.resample(order=2, events, synthesized_stim=True)
then the events could get resampled, and you could add a new supplemental STIM channel with that information.
I just wanted this discussion as part of the design process.

wmvanvliet commented:
It's good to have this discussion.

agramfort commented:
LGTM but it should be compared to C code strategy

wmvanvliet commented:
This is the approach taken in `mne_browse_raw`:

Say we wish to decimate the stim channel by a factor of 10.

 1. Split the stim channel data in windows of length 10.
 2. For each window, take the first non-zero value.

This approach is simple and covers some of the more obvious problems. However, it will still silently drop events if there are two events occurring in the same window. Also, two events can be merged as one.


    `y = [0, 0, 0, 1, 0, 1, 0, 0]`

    `y = [0, 0, 1, 1, 0]`

so now two events become one... whoops. (The code in this PR would throw an error on this example, which is preferable to silently messing up the events).

wmvanvliet commented:
How about this for an API?

    events = raw.find_events()
    raw.resample(order=2, events)  # raw and events are modified in-place

I don't really see a point in reconstructing the stim channel if events have already been extracted.

wmvanvliet commented:
cc @agramfort 

Unfortunately, there is no straightforward way to reconstruct a stim channel based on an event matrix:

 1. We don't know what stimulus channel the events were created from. Could be one. Could be many.
 2. We support extracting events based on onsets/offsets/step and other variations.

Theoretically, this could be solved, but it would make the code much more complicated than it is now.

dgwakeman commented:
First, that was just an added feature I thought of to enhance the flexibility of the API I mentioned. The main advantage of the API is that it demands feeding events to resample, so the user can check that all of the events are caught, plus, we can issue errors then if we know they will lose events with their settings e.g. ```[0 1 0 2 0 1 0 2 01 ]``` and a decimation factor of 10 (as you mentioned above). You must lose data here, and we can issue an error. 
Point 2 makes sense, in that the reconstructed stimulus channel won't be a "perfect" downsample of the stim channel, but it would contain what people were interested in in the beginning. Again, I don't think this feature is necessary.
I do see a huge advantage to making a requirement to resample be the events structure, because then at least the user has to think about what sort of events they want to get out of their resampled data instead of just assuming it worked and then getting annoyed when things break. @agramfort thoughts

agramfort commented:
I proposed a fix to @wmvanvliet 

Let's see if you agree on the outcome. Hopefully soon

jona-sassenhagen commented:
I'd really appreciate a fix for this, it's a hidden failure scenario I've experienced a few times. +1 for this.

wmvanvliet commented:
Ok, this PR now aligns how stim channels are resampled with the way `mne_browse_raw` does it.

In addition, I've implemented a new API.
Now you can do this:

    events = mne.find_events(raw)
    raw.resample(100, events)
    epochs = mne.Epochs(raw, events, ...)

By giving the event matrix to the resample method, the data and event onsets can be jointly resampled, bypassing the problematic resampling of the stim channel completely. In this instance the raw data and events matrix are modified in-place. I've also added a `copy` parameter to the resample method, so you could also do this:
    
    events = mne.find_events(raw)
    raw_res, events_res = raw.resample(100, events, copy=True)
    epochs = mne.Epochs(raw_res, events_res, ...)
    
Giving the `events` to the `resample` function is of course optional. Backwards compatibility is preserved.
I've added some checks that warns the user if the resampling of the stim channels lead to problems.

Eric89GXL commented:
That is a clever solution!

wmvanvliet commented:
I'm happy with the API now. This is ready for review and hopefully merge.

Eric89GXL commented on examples/preprocessing/plot_resample.py:
```diff
@@ -0,0 +1,55 @@
+print('Number of events after resampling:', len(events_resampled))
+
+###############################################################################
+# Epoched data can be resampled as well.
```
add note that this can effectively jitter event timings

wmvanvliet replied:
you lose temporal resolution both on the data and the events. Instead of having a resolution of 600 Hz they both now have a resolution of 100 Hz. This is self evident is it not?

wmvanvliet replied:
event timing receive no extra jitter beyond this loss in resolution.

wmvanvliet replied:
meh, people can be stupid. I will make a note of it :)

Eric89GXL replied:
the thing is, you *don't* get the jitter if you resample epochs

Eric89GXL replied:
See this `gist` that hopefully makes it clear:

https://gist.github.com/Eric89GXL/01642cb3789992fbca59

Eric89GXL replied:
From a signal-processing perspective, what epoching *before* downsampling gets you is that you get to pull out data from the original data in what would end up being fractional sample locations of the downsampled raw signal. You could get something equivalent by downsampling your epochs, and then correcting for how far off you were by introducing a sub-sample delay (assuming you took care of the edges appropriately). So epoching then downsampling really is a better solution from a phase alignment perspective.

Eric89GXL replied:
Think of a signal that you want to downsample by a factor of 10, but has an "event" at sample number 5 in the original data. Epoching first, your "sample zero" will correctly be at sample number 5, and then you will decimate/downsample that epoch, meaning the first sample of that epoch comes from what was originally sample number 5.

If you downsample then epoch, your "sample zero" at the new rate will have to come from sample number 0 or 10 at the original rate (because these are the only samples you have access to now), so you've introduced a jitter of 5 samples (at the original rate) or half a sample (at the new rate) that wouldn't be there in the other regime. Make sense?

wmvanvliet replied:
point taken. I'll modify the text of the example.

Eric89GXL commented on mne/filter.py:
```diff
@@ -1346,6 +1346,55 @@ def resample(x, up, down, npad=100, axis=-1, window='boxcar', n_jobs=1,
+    stim_resampled : 2D array, shape (n_stim_channels, n_samples_resampled)
+        The resampled stim channels
+    """
+    stim_data = np.atleast_2d(stim_data)
```
Can you make a note of where the equivalent steps are done in the C code? We try to make it easier for the next person to look up code if possible. Did you end up making something equivalent, or just close? Please add a note for that, too (also good to know where we differ, if at all).

Eric89GXL commented on mne/io/base.py:
```diff
@@ -926,8 +927,8 @@ def notch_filter(self, freqs, picks=None, filter_length='10s',
+                 n_jobs=1, events=None, copy=False, verbose=None):
         """Resample data channels.
 
         Resamples all channels. The data of the Raw object is modified inplace.
```
the inplace-ness is not necessarily true anymore, update docstring

Eric89GXL commented on mne/io/base.py:
```diff
@@ -938,9 +939,10 @@ def resample(self, sfreq, npad=100, window='boxcar',
+                     generally recommended to find events using the original
+                     data and jointly resample the raw and events using the
+                     'events' parameter. Alternatively, resampling can be
+                     performed on the epochs instead. See here for an
```
Please restore the original warning suggesting use of `epochs` resampling over that of `raw`. Even when resampling the events there will be effective jitter.

agramfort commented on examples/preprocessing/plot_resample.py:
```diff
@@ -0,0 +1,63 @@
+
+
+Resampling can save memory and computation time, but comes at a cost of a loss
+in temporal solution.
```
solution -> resolution

agramfort commented on examples/preprocessing/plot_resample.py:
```diff
@@ -0,0 +1,63 @@
+Resampling can save memory and computation time, but comes at a cost of a loss
+in temporal solution.
+This achieves a 6-fold reduction in data size, at the cost of an equal loss of
+temporal resolution.
```
dupe: "", at the cost of an equal loss of
+temporal resolution.

jona-sassenhagen replied:
slow**s** down computations

dgwakeman commented:
Glad to see the events API added! 
I'm +1/-1 on the following idea, but I'll let you think about it:
renaming/removing the stim channel (at least when events is included in the call). This way people don't forget they did it. (+1 for safety, -1 for they should know what they did to their data). 

Not for this PR: Also, what do people think about the idea of adding `events` to `info`, to decrease manual book keeping?

wmvanvliet commented:
actually, `info` **has** a field for events. I came across it while documenting the beast. It's not used though.

wmvanvliet commented:
> renaming/removing the stim channel 

-1 for me. It makes things more complex and less transparent. I think the multitude of warnings in the manual, example and docstring should be enough.

Eric89GXL commented:
-1 on adding `events` to info, I think it would be more confusing than helpful. If it's just going to reflect what is in the stim channel, now we have violated DRY by making us keep track in multiple places all the time. If it's supposed to be something more, I don't know if it's worth it. We can't really know what events people plan on using. In our pipelines at UW, we use multiple event lists, so having one that is part of `info` wouldn't really make sense... I imagine other cases are similar.

dgwakeman commented:
sounds good!

jona-sassenhagen commented:
IMO it would be EXTREMELY useful to have a field containing per-epoch events information that is updated together with the epochs object (such as response time on the corresponding trial). So e.g. when epochs are dropped, the corresponding events are also dropped. This is done in EEGLAB, and would really simplify a bunch of stuff that's trivial in EEGLAB, such as basically all the single-trial stuff I did in my Brain & Language and Cortex papers.

But that's different from keeping the raw events array itself.

wmvanvliet commented:
@jona-sassenhagen we could add numerical `id` numbers to the epochs. By default this would just be `range(len(epochs))`, but as epochs are dropped (bad epoch detection and whatnot) the corresponding ids are removed from the list. This allows you, for example, to easily keep a pandas dataframe aligned with the epochs object.

jona-sassenhagen commented:
I'd love that, but that's really not for this PR. Finish it up first, it'd be super helpful!

Eric89GXL commented:
Don't we basically already have that with `epochs.selection`?

wmvanvliet commented:
@Eric89GXL you are right! We have it! good news @jona-sassenhagen :)

jona-sassenhagen commented:
Oh, that reminds me why just having integer ids is not so cool, and some arbitrary pointer would be nicer: if you e.g. merge epoch files, you have to manually take care of things again, which can get very messy (at least the way I did it). It would be nicer if that was done automatically.

wmvanvliet commented:
Ok, I've finished processing your comments. Are we ready for merge?

Eric89GXL commented on mne/io/base.py:
```diff
@@ -940,11 +941,15 @@ def resample(self, sfreq, npad=100, window='boxcar',
 
+                     Aternatively, events can be constructed using the original
+                     data and jointly resampled with the raw data using the
+                     'events' parameter.
```
Sorry to be a pest about this, but saying `Alternatively` after the warning makes it sound like this method will fix the issue, but it won't. So please move elsewhere (and fix `Aternatively` spelling)

agramfort commented on examples/preprocessing/plot_resample.py:
```diff
@@ -0,0 +1,84 @@
+import mne
+from mne.io import Raw
+from mne.datasets import sample
+from matplotlib import pyplot as plt
```
import before mne

agramfort commented on examples/preprocessing/plot_resample.py:
```diff
@@ -0,0 +1,84 @@
+# Since downsampling reduces the timing precision of events, we recommend
+# first extracting epochs and downsampling the Epochs object:
+events = mne.find_events(raw)
+epochs = mne.Epochs(raw, events, [1, 2, 3, 4], tmin=-0.1, tmax=0.8,
```
please name the param event_id

agramfort commented on examples/preprocessing/plot_resample.py:
```diff
@@ -0,0 +1,84 @@
+plt.xlabel('time (s)')
+plt.legend(['original', 'downsampled'], loc='best')
+plt.title('Effect of downsampling')
+plt.tight_layout()
```
mne.viz.utils.tight_layout

agramfort commented on mne/filter.py:
```diff
@@ -1346,6 +1346,60 @@ def resample(x, up, down, npad=100, axis=-1, window='boxcar', n_jobs=1,
+    The approach taken here is equivalent to the approach in the C-code.
+    See the decimate_stimch function in MNE/mne_browse_raw/save.c
+    """
+    stim_data = np.atleast_2d(stim_data)
```
if you do atleast_2d it means you accept 1d data. The doc says only 2d

agramfort commented on mne/io/base.py:
```diff
@@ -973,45 +990,72 @@ def resample(self, sfreq, npad=100, window='boxcar',
+
+            events[:, 0] = np.minimum(
+                np.round(events[:, 0] * ratio).astype(int),
+                inst._data.shape[1]
```
why is the np.minimum needed here?

Eric89GXL replied:
the round could move you outside the bounds I think

in terms of not being just `min` it's because it's an array that's being operated on

wmvanvliet replied:
@Eric89GXL is right. It is a boundary check.

agramfort replied:
ok fair enough

agramfort commented on mne/io/fiff/tests/test_raw.py:
```diff
@@ -821,6 +821,59 @@ def test_resample():
+    # basic decimation
+    stim = [1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0]
+    raw = RawArray([stim], create_info(1, len(stim), ['stim']))
+    assert_allclose(raw.resample(8)._data,
```
8 -> 8.

it's a float

wmvanvliet replied:
8 is not a float. 8 is an integer.

`raw.resample` should be able to deal with integers.

agramfort commented on mne/io/fiff/tests/test_raw.py:
```diff
@@ -821,6 +821,59 @@ def test_resample():
+                    [[1, 1, 0, 0, 1, 1, 0, 0]])
+
+    # decimation of multiple stim channels
+    raw = RawArray(2*[stim], create_info(2, len(stim), 2*['stim']))
```
spaces around * pep8

agramfort commented on mne/io/fiff/tests/test_raw.py:
```diff
@@ -821,6 +821,59 @@ def test_resample():
+
+    # decimation of multiple stim channels
+    raw = RawArray(2*[stim], create_info(2, len(stim), 2*['stim']))
+    assert_allclose(raw.resample(8)._data,
```
8.

agramfort commented on mne/io/fiff/tests/test_raw.py:
```diff
@@ -821,6 +821,59 @@ def test_resample():
+    # done naively
+    stim = [0, 0, 0, 1, 1, 0, 0, 0]
+    raw = RawArray([stim], create_info(1, len(stim), ['stim']))
+    assert_allclose(raw.resample(4)._data,
```
4.

see also below

agramfort commented:
besides LGTM !

agramfort commented:
>
> Not for this PR: Also, what do people think about the idea of adding
> events to info, to decrease manual book keeping?
>
events is not a measurement info thing.

agramfort commented:
>
> Oh, that reminds me why just having integer ids is not so cool, and some
> arbitrary pointer would be nicer: if you e.g. merge epoch files, you have
> to manually take care of things again, which can get very messy (at least
> the way I did it). It would be nicer if that was done automatically.
>

I think epochs have all the info you need to do what you're asking.
Paste a "messy" code snippet so we can evaluate how lazy you are :)

agramfort commented:
besides +1 for merge

thanks @wmvanvliet !

wmvanvliet commented:
Travis is finally happy!

@agramfort Can you do a final check before I merge this? The version you last checked contained a flaw where resampling wouldn't work when no status channel was present. This is solved now with an `except: pass` statement in one of the very rare cases I believe it is appropriate.

agramfort commented:
LGTM

please rebase, update what's new and merge.

thanks @wmvanvliet !

agramfort commented:
:beer: @wmvanvliet  !

jona-sassenhagen commented:
nice
